### PR TITLE
Preface, Preliminaries, and RNN Full Scan

### PR DESCRIPTION
**Pull request by goldmermaid**
1. Code:
    1. Import in alphabetical order
    2. 78 chars per line with ruler
    3. Lint
    4. Remove period from comments

2. Text:
    1. Title capitalization & Verb to noun
    2. Figure referring
    3. Spell typo fixing

### CODE REVIEW

File: chapter_recurrent-neural-networks/bi-rnn.md [hunk 1]
```diff
@@ -63,18 +64,17 @@ with initialization $\rho_T(h_T) = 1$. These two recursions allow us to sum over
...
+Note that in abstract terms the backward recursion can be written as $\rho_{t-1} = g(\rho_t, x_t)$, where $g$ is a learnable function. Again, this looks very much like an update equation, just running backwards unlike what we have seen so far in RNNs. Indeed, HMMs benefit from knowing future data when it is available. Signal processing scientists distinguish between the two cases of knowing and not knowing future observations as interpolation v.s. exploration. See the introductory chapter of the book by :cite:`Doucet.De-Freitas.Gordon.2001` on sequential Monte Carlo algorithms for more details.
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/bi-rnn.md]: mytodo: extrapolation

File: chapter_recurrent-neural-networks/machine-translation.md [hunk 2]
```diff
@@ -113,7 +116,7 @@ def build_array(lines, vocab, num_steps, is_source):
 
 Then we can construct minibatches based on these arrays. 
 
-## Put all Things Together
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/machine-translation.md]: mytodo, putting all

File: chapter_recurrent-neural-networks/rnn-scratch.md [hunk 3]
```diff
@@ -220,12 +222,14 @@ def train_ch8(model, train_iter, vocab, lr, num_epochs, ctx,
                             legend=['train'], xlim=[1, num_epochs])
     if isinstance(model, gluon.Block):
         model.initialize(ctx=ctx, force_reinit=True, init=init.Normal(0.01))
-        trainer = gluon.Trainer(model.collect_params(), 'sgd', {'learning_rate': lr})
-        updater = lambda batch_size : trainer.step(batch_size)
+        trainer = gluon.Trainer(model.collect_params(),
+                                'sgd', {'learning_rate': lr})
+
+        def updater(batch_size): return trainer.step(batch_size)
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/rnn-scratch.md]: mytodo, newline

File: chapter_recurrent-neural-networks/rnn-scratch.md [hunk 4]
```diff
@@ -220,12 +222,14 @@ def train_ch8(model, train_iter, vocab, lr, num_epochs, ctx,
...
     else:
-        updater = lambda batch_size : d2l.sgd(model.params, lr, batch_size)
+        def updater(batch_size): return d2l.sgd(model.params, lr, batch_size)
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/rnn-scratch.md]: mytodo, newline

File: chapter_recurrent-neural-networks/rnn-scratch.md [hunk 5]
```diff
@@ -220,12 +222,14 @@ def train_ch8(model, train_iter, vocab, lr, num_epochs, ctx,
...
 
-    predict = lambda prefix: predict_ch8(prefix, 50, model, vocab, ctx)
+    def predict(prefix): return predict_ch8(prefix, 50, model, vocab, ctx)
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/rnn-scratch.md]: mytodo, newline

File: chapter_recurrent-neural-networks/seq2seq.md [hunk 6]
```diff
@@ -104,7 +107,8 @@ class Seq2SeqDecoder(d2l.Decoder):
     def forward(self, X, state):
         X = self.embedding(X).swapaxes(0, 1)
         out, state = self.rnn(X, state)
-        # Make the batch to be the first dimension to simplify loss computation.
+        # Make the batch to be the first dimension to
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/seq2seq.md]: mytodo: be greedy in 78 chars

File: chapter_recurrent-neural-networks/text-preprocessing.md [hunk 7]
```diff
@@ -104,11 +105,11 @@ After that, we can convert each sentence into a list of numerical indices. To il
 
 ```{.python .input  n=25}
 for i in range(8, 10):
-    print('words:', tokens[i]) 
+    print('words:', tokens[i])
     print('indices:', vocab[tokens[i]])
 ```
 
-## Put All Things Together
+## Putting Things Together
```
Comment (astonzhang) [inline chapter_recurrent-neural-networks/text-preprocessing.md]: mytodo

File: chapter_recurrent-neural-networks/bi-rnn.md (hunk 1 above)
Comment (goldmermaid) [reply chapter_recurrent-neural-networks/bi-rnn.md]: Thanks!

File: chapter_recurrent-neural-networks/machine-translation.md (hunk 2 above)
Comment (goldmermaid) [reply chapter_recurrent-neural-networks/machine-translation.md]: Oh...

**Pull request merged**